# Bài tập tuần 7

## Thông tin học viên

| Họ và tên |MSHV |
|-----------|----------------|
|Trần Hoàng Quân|19120338|

## Import các thư viện cần thiết

In [1]:
import numpy as np
import scipy.stats as stats
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
from patsy import dmatrices

# Đề bài

Mô hình hồi qui giá nhà ($PRICE$ tính bằng *nghìn đô*) phụ thuộc vào các biến $SQFT$ (diện tích nhà), $BEDRMS$ (số phòng ngủ) và $BATHS$ (số phòng tắm)

|$i$|$PRICE$|$SQFT$|$BEDRMS$|$BATHS$|
|-|-----|----|------|-----|
|1|199.9|1065|3|1.75|
|2|228|1254|3|2|
|3|235|1300|3|2|
|4|285|1577|4|2.5|
|5|239|1600|3|2|
|6|293|1750|4|2|
|7|285|1800|4|2.75|
|8|365|1870|4|2|
|9|295|1935|4|2.5|
|10|290|1948|4|2|
|11|385|2254|4|3|
|12|505|2600|3|2.5|
|13|425|2800|4|3|
|14|415|3000|4|3|

1. Hồi qui mẫu
   $$
     PRICE_i = \beta_0 + SQFT_i\beta_1 + BEDRMS_i\beta_2 + BATHS_i\beta_3 + \epsilon_i
   $$
   và giải thích ý nghĩa các hệ số hồi qui?
2. Tính ước lượng điểm giá $PRICE$ trung bình khi $SQFT$ tăng thêm $300$ và $BEDRMS$ tăng thêm $1$
3. Tính hệ số xác định và hệ số xác định điều chỉnh, giải thích ý nghĩa?
4. Tìm khoảng tin cậy đối với $\beta_1, \beta_2$ và $\beta_3$ với mức ý nghĩa $5\%$
5. Phải chăng cả ba biến đều không ảnh hưởng đến giá trị $PRICE$?
6. Dự báo giá trị trung bình và giá trị cá biệt về giá nhà trung bình khi $SQRT = 2500$, $BEDRMS = 3$ và $BATHS = 2$? 

# Câu 1

## Hồi qui mẫu $PRICE_i = \beta_0 + SQFT_i\beta_1 + BEDRMS_i\beta_2 + BATHS_i\beta_3 + \epsilon_i$

In [2]:
df = pd.DataFrame({
    "price": [199.9, 228, 235, 285, 239, 293, 285, 365, 295, 290, 385, 505, 425, 415],
    "sqft": [1065, 1254, 1300, 1577, 1600, 1750, 1800, 1870, 1935, 1948, 2254, 2600, 2800, 3000],
    "bedrms": [3, 3, 3, 4, 3, 4, 4, 4, 4, 4, 4, 3, 4, 4],
    "baths": [1.75, 2, 2, 2.5, 2, 2, 2.75, 2, 2.5, 2, 3, 2.5, 3, 3]
})

In [3]:
df

price  sqft  bedrms  baths
0   199.9  1065       3   1.75
1   228.0  1254       3   2.00
2   235.0  1300       3   2.00
3   285.0  1577       4   2.50
4   239.0  1600       3   2.00
5   293.0  1750       4   2.00
6   285.0  1800       4   2.75
7   365.0  1870       4   2.00
8   295.0  1935       4   2.50
9   290.0  1948       4   2.00
10  385.0  2254       4   3.00
11  505.0  2600       3   2.50
12  425.0  2800       4   3.00
13  415.0  3000       4   3.00

In [4]:
regression = smf.ols("price ~ sqft + bedrms + baths", data = df).fit()

In [5]:
print(regression.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.836
Model:                            OLS   Adj. R-squared:                  0.787
Method:                 Least Squares   F-statistic:                     16.99
Date:                Wed, 14 Jun 2023   Prob (F-statistic):           0.000299
Time:                        11:57:17   Log-Likelihood:                -69.454
No. Observations:                  14   AIC:                             146.9
Df Residuals:                      10   BIC:                             149.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    129.0616     88.303      1.462      0.1

D:\Software\Python\lib\site-packages\scipy\stats\_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=14
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Phương trình hồi qui tuyến tính khi đó có dạng

$$
PRICE = 129.0616 + (0.1548\times SQFT) - (21.5875\times BEDRMS) - (12.1928\times BATHS)
$$

với

- $SQFT$: diện tích nhà
- $BEDRMS$: số lượng phòng ngủ
- $BATHS$: số lượng phòng tắm

## Ý nghĩa các hệ số hồi qui?

Xét hàm hồi quy 

$$
PRICE = 129.0616 + (0.1548\times SQFT) - (21.5875\times BEDRMS) - (12.1928\times BATHS)
$$

với $\beta_0 = 129.0616, \beta_1 = 0.1548, \beta_2 = -21.5875, \beta_3 = -12.1928$. Ý nghĩa của các hệ số hồi qui lần lượt như sau:

- $\beta_0 = 129.0616$: trong điều kiện các yếu tố khác không đổi, khi căn nhà có diện tích là $0$ (sqrf), $0$ phòng ngủ và $0$ nhà tắm thì giá của căn nhà là $129.0616$ (nghìn USD)
- $\beta_1 = 0.1548$: trong điều kiện các yếu tố khác không đổi, khi tăng (giảm) diện tích căn nhà lên $1$ sqrf thì giá căn nhà sẽ tăng (giảm) thêm $0.1548$ (nghìn USD)
- $\beta_2 = -21.5875$: trong điều kiện các yếu tố khác không đổi, khi tăng (giảm) số lượng phòng ngủ thêm $1$ phòng thì giá căn nhà sẽ giảm (tăng) thêm $21.5875$ (nghìn USD)
- $\beta_3 = -12.1928$: trong điều kiện các yếu tố khác không đổi, khi tăng (giảm) số lượng phòng tắm thêm $1$ phòng thì giá căn nhà sẽ giảm (tăng) thêm $12.1928$ (nghìn USD)

# Câu 2

## Tính ước lượng điểm giá $PRICE$ trung bình khi $SQFT$ tăng thêm $300$ và $BEDRMS$ tăng thêm $1$

Khi đó công thức hồi qui có dạng

$$
PRICE_{new} = 129.0616 + [0.1548\times (SQFT + 300)] - [21.5875\times (BEDRMS + 1)] - (12.1928\times BATHS)
$$

In [6]:
df["sqft_new"] = df["sqft"] + 300
df["bedrms_new"] = df["bedrms"] + 1

In [7]:
df["price_new"] = 129.0616 + 0.1548*df["sqft_new"] - 21.5875 * df["bedrms_new"] - 12.1928 * df["baths"]

In [8]:
df

price  sqft  bedrms  baths  sqft_new  bedrms_new  price_new
0   199.9  1065       3   1.75      1365           4   232.6762
1   228.0  1254       3   2.00      1554           4   258.8852
2   235.0  1300       3   2.00      1600           4   266.0060
3   285.0  1577       4   2.50      1877           5   281.2017
4   239.0  1600       3   2.00      1900           4   312.4460
5   293.0  1750       4   2.00      2050           5   314.0785
6   285.0  1800       4   2.75      2100           5   312.6739
7   365.0  1870       4   2.00      2170           5   332.6545
8   295.0  1935       4   2.50      2235           5   336.6201
9   290.0  1948       4   2.00      2248           5   344.7289
10  385.0  2254       4   3.00      2554           5   379.9049
11  505.0  2600       3   2.50      2900           4   461.1496
12  425.0  2800       4   3.00      3100           5   464.4257
13  415.0  3000       4   3.00      3300           5   495.3857

In [9]:
mean_price_new = df["price_new"].mean()
print(f"{mean_price_new:.4f}")

342.3455


Vậy, khi $SQFT$ tăng thêm $300$ (sqft) và $BEDRMS$ tăng thêm $1$ (phòng) thì ước lượng điểm giá $PRICE$ trung bình $\mu_{\hat{PRICE_{new}}} = 342.3455$ (nghìn đô).

# Câu 3

## Tính hệ số xác định và hệ số xác định điều chỉnh?

Hệ số xác định $R^2$ và hệ số xác định điều chỉnh $R^2_{adj}$ được tính theo công thức như sau:

$$
\displaystyle
TSS = \sum_{i = 1}^{n} (y_i - \bar{y})^2
$$
$$
\displaystyle
ESS = \sum_{i = 1}^{n} (\hat{y}_i - \bar{y})^2
$$
$$
\displaystyle
R^2 = \frac{ESS}{TSS}
$$
$$
\displaystyle
R^2_{adj} = 1 - (1 - R^2)\frac{n - 1}{n - k - 1}
$$

với $k$ là số biến độc lập trong phương trình hồi quy tuyến tính.

In [10]:
y_mean = df["price"].mean()
y_hat = (129.0616 + 0.1548*df["sqft"] - 21.5875 * df["bedrms"] - 12.1928 * df["baths"]).to_numpy()
y = df["price"].to_numpy()

In [11]:
y_mean, y_hat, y

(317.49285714285713,
 array([207.8237, 234.0327, 241.1535, 256.3492, 287.5935, 289.226 ,
        287.8214, 307.802 , 311.7676, 319.8764, 355.0524, 436.2971,
        439.5732, 470.5332]),
 array([199.9, 228. , 235. , 285. , 239. , 293. , 285. , 365. , 295. ,
        290. , 385. , 505. , 425. , 415. ]))

In [12]:
n = df.shape[0]
k = 3

In [13]:
tss = np.sum((y - y_mean) ** 2)
ess = np.sum((y_hat - y_mean) ** 2)
r_squared = ess / tss
r_squared_adj = 1 - (1 - r_squared) * ((n - 1) / (n - k - 1))

In [14]:
print(f"r_squared = {r_squared:.4f}")
print(f"r_squared_adj = {r_squared_adj:.4f}")

r_squared = 0.8360
r_squared_adj = 0.7868


## Giải thích ý nghĩa?

Hệ số xác định $R^2_{adj} = 0.7868$ thể hiện mô hình hồi qui có thể đưa ra giải thích cho khoảng $78,68\%$ giá nhà $PRICE$ dựa trên các biến kích thước nhà $SQFT$, số phòng ngủ $BEDRMS$ và số phòng tắm $BATHS$.

# Câu 4

Với mức ý nghĩa $\alpha$, khoảng tin cậy $CI_{\alpha}^{\beta_i}$ cho tham số $\beta_i$ được tính bằng công thức:

$$
CI_{\alpha}^{\beta_i} = \left[\beta_i -  SE(\beta_i)\times t\left(1 - \frac{\alpha}{2}, n - 2\right), \beta_i + SE(\beta_i)\times t\left(1 - \frac{\alpha}{2}, n - 2\right)\right]
$$

với $\alpha$ là mức ý nghĩa, $n$ là số lượng mẫu. $SE(\beta_i)$ được sử dụng lại từ câu 1.

## Tìm khoảng tin cậy cho $\beta_1$ với mức ý nghĩa $5\%$

In [15]:
alpha = 0.05

Với $\beta_1 = 0.1548$, ta có $SE(\beta_1) = 0.032$, $\alpha = 0.05$, $n = 14$. Khi đó $CI_{\alpha}^{\beta_i}$ được tính như sau:


In [16]:
p_value_beta_1 = stats.t.ppf(1 - alpha/2, df.shape[0] - 2)
se_beta_1 = 0.032
beta_1 = 0.1548

In [17]:
p_value_beta_1

2.1788128296634177

In [18]:
print(f"ci_beta_1 = [{(beta_1 - se_beta_1 * p_value_beta_1):.4f}, {(beta_1 + se_beta_1 * p_value_beta_1):.4f}]")

ci_beta_1 = [0.0851, 0.2245]


Vậy, khoảng tin cậy cho $\beta_1$ với mức ý nghĩa $5\%$ là $[0.0851, 0.2245]$

## Tìm khoảng tin cậy cho $\beta_2$ với mức ý nghĩa $5\%$

Với $\beta_2 = -21.5875$, ta có $SE(\beta_2) = 27.029$, $\alpha = 0.05$, $n = 14$. Khi đó $CI_{\alpha}^{\beta_i}$ được tính như sau:


In [19]:
p_value_beta_2 = stats.t.ppf(1 - alpha/2, df.shape[0] - 2)
se_beta_2 = 27.029
beta_2 = -21.5875

In [20]:
p_value_beta_2

2.1788128296634177

In [21]:
print(f"ci_beta_2 = [{(beta_2 - se_beta_2 * p_value_beta_2):.4f}, {(beta_2 + se_beta_2 * p_value_beta_2):.4f}]")

ci_beta_2 = [-80.4786, 37.3036]


Vậy, khoảng tin cậy cho $\beta_2$ với mức ý nghĩa $5\%$ là $[-80.4786, 37.3036]$

## Tìm khoảng tin cậy cho $\beta_3$ với mức ý nghĩa $5\%$

Với $\beta_3 = -12.1928$, ta có $SE(\beta_3) = 43.250$, $\alpha = 0.05$, $n = 14$. Khi đó $CI_{\alpha}^{\beta_i}$ được tính như sau:


In [22]:
p_value_beta_3 = stats.t.ppf(1 - alpha/2, df.shape[0] - 2)
se_beta_3 = 43.250
beta_3 = -12.1928

In [23]:
p_value_beta_3

2.1788128296634177

In [24]:
print(f"ci_beta_1 = [{(beta_3 - se_beta_3 * p_value_beta_3):.4f}, {(beta_3 + se_beta_3 * p_value_beta_3):.4f}]")

ci_beta_1 = [-106.4265, 82.0409]


Vậy, khoảng tin cậy cho $\beta_3$ với mức ý nghĩa $5\%$ là $[-106.4265, 82.0409]$

# Câu 5

## Phải chăng cả ba biến đều không ảnh hưởng đến giá trị $PRICE$?

Sử dụng lại kết quả từ câu 1:

In [25]:
print(regression.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.836
Model:                            OLS   Adj. R-squared:                  0.787
Method:                 Least Squares   F-statistic:                     16.99
Date:                Wed, 14 Jun 2023   Prob (F-statistic):           0.000299
Time:                        11:57:18   Log-Likelihood:                -69.454
No. Observations:                  14   AIC:                             146.9
Df Residuals:                      10   BIC:                             149.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    129.0616     88.303      1.462      0.1

D:\Software\Python\lib\site-packages\scipy\stats\_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=14
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Ta có thể nhận thấy các giá trị $P > |t|$ của các biến $BEDRMS$ và $BATHS$ lần lượt là $0.443$ và $0.784$. Đây là dấu hiệu cho thấy các biến $BEDRMS$ và $BATHS$ không có ý nghĩa về mặt thống kê và có thể loại bỏ.

Vậy, chỉ có các biến $BEDRMS$ và $BATHS$ không ảnh hưởng đến giá nhà $PRICE$.

# Câu 6

## Dự báo giá trị trung bình và giá trị cá biệt về giá nhà trung bình khi $SQFT = 2500, BEDRMS = 3$ và $BATHS = 2$?